### PR TITLE
Allow maximum number of domains to be specified as a OCAMLRUNPARAM parameter

### LINKED ISSUE #13256: Raise the maximum number of domains
It seems like currently OCaml cannot create more than 128 domains (creating more fails with `Fatal error: exception Failure("failed to allocate domain")`). I am trying to run a parallel OCaml program on a Google Cloud VM with 180 cores and 360 vCPUs (AMD Genoa C3D). Is it possible to raise this limit, ideally to at least 360? Thanks!

### PR DESCRIPTION
**Pull request by kayceesrk**
This PR introduces a new `OCAMLRUNPARAM` parameter `d` that permits the user to set the maximum number of domains. This idea was suggested by @shindere at https://github.com/ocaml/ocaml/issues/13256#issuecomment-2182422907. 

Fixes #13256.

### CODE REVIEW

**Comment by ghost** [comment]
Shouldn't the user-configurable value be bounds checked to be within reasonable limits (i.e. 1 <= user value < no more than a few k)?

**Comment by kayceesrk** [comment]
> Shouldn't the user-configurable value be bounds checked to be within reasonable limits (i.e. 1 <= user value < no more than a few k)?

That sounds reasonable. Would 4096 be a good upper limit for a start?

**Comment by gasche** [comment]
I believe that this is a step in the right direction -- hardcoding a fixed maximum of domains is bound to create issues sooner or later.

I tried to use github search to find code out there that relies on the current limit to the number of domains. It's hard because there is a lot of noise from copies of the compiler codebase. I found the following:

- [boxroot](https://gitlab.com/ocaml-rust/ocaml-boxroot/), as found in a vendored copy at https://github.com/lord/ocaml-swift-example/blob/1fd0033c24d54ff4ae962c59b59a10594c2268c6/boxroot/platform.c#L11-L13
- multicoretests in https://github.com/shym/multicoretests/blob/22e3440e3158efeb52d5418b242207e100c3b6a2/src/domain/domain_spawntree.ml#L3 ; this seems slightly harder to fix as it runs as a separate process, and does not have access to the OCAMLRUNPARAM value

Boxroot is easily fixed by reading the new global instead, I believe. (This code runs inside the OCaml process, so it sees the same OCAMLRUNPARAM value). For multicoretests this is less obvious because it runs as a separate process, so it cannot access the "real" value, but in fact this is just fine as long as the test runner controls the OCaml process and can fix the max_domains value there.

For runtime_events consumers we should be in the clear, as the runtime_events codebase had the excellent insight of passing the `max_domains` value in the metadata header from the start: it is already possible to write consumers that are robust to different values of this.

To summarize: I think that we are making this change early enough that there is not a lot of sensitivity on this value in the ecosystem, and we should be fine compatibility-wise.

**Comment by ghost** [comment]
> That sounds reasonable. Would 4096 be a good upper limit for a start?

I think so.

Or maybe `2 x sysconf(_SC_NPROCESSORS_CONF)` on Unix-like platforms. But that's probably too much bikeshed already.

**Comment by shindere** [comment]
KC Sivaramakrishnan (2024/07/01 02:54 -0700):
> > Shouldn't the user-configurable value be bounds checked to be within reasonable limits (i.e. 1 <= user value < no more than a few k)?
>
> That sounds reasonable. Would 4096 be a good upper limit for a start?

I have no idea / experience but are there values that would just make no
sense in terms of cache sizes or so?

**Comment by gasche** [comment]
> Or maybe `2 x sysconf(_SC_NPROCESSORS_CONF)` on Unix-like platforms. But that's probably too much bikeshed already.

One issue with this is that it would break on the current default for most machines, so for example if I run the multicoretest runner with `OCAMLRUNPARAM="domains=128"` I get a hard failure.

@kayceesrk why are we using one-letter name for OCAMLRUNPARAM variables, is this enforced by the specification or implementation? Otherwise I wouldn't mind `domains` or `max_domains` instead of `d`.

**Comment by shindere** [comment]
Gabriel Scherer (2024/07/01 02:59 -0700):
> > Or maybe `2 x sysconf(_SC_NPROCESSORS_CONF)` on Unix-like platforms. But that's probably too much bikeshed already.
>
> One issue with this is that it would break on the current default for most machines, so for example if I run the multicoretest runner with `OCAMLRUNPARAM="domains=128"` I get a hard failure.
>
> @kayceesrk why are we using one-letter name for OCAMLRUNPARAM
> variables, is this enforced by the specification or implementation?
> Otherwise I wouldn't mind `domains` or `max_domains` instead of `d`.

@damiendoligez may have something to say on that one?

**Comment by kayceesrk** [comment]
> For runtime_events consumers we should be in the clear, as the runtime_events codebase had the excellent insight of passing the max_domains value in the metadata header from the start: it is already possible to write consumers that are robust to different values of this.

Thanks for writing this out. I did want to mention this in the PR message. There was nothing to be done here since @sadiqj has already done the necessary work. Thanks Sadiq!

**Comment by kayceesrk** [comment]
> To summarize: I think that we are making this change early enough that there is not a lot of sensitivity on this value in the ecosystem, and we should be fine compatibility-wise.

Thanks for the summary. 

To fix the failing windows builds, @dra27 mentioned that this may need a bootstrap because of flexdll bootstrap. I'll do that next.

File: manual/src/cmds/runtime.etex [hunk 1]
```diff
@@ -129,6 +129,8 @@ The following environment variables are also consulted:
         "caml_shutdown" in section~\ref{ss:c-embedded-code}). The option also enables
         pooling (as in "caml_startup_pooled"). This mode can be used to detect
         leaks with a third-party memory debugger.
+  \item[d] ("max_domains") Maximum number of domains that can be active
+        concurrently. Defaults to 128 in native and 16 in bytecode runtimes.
```
Comment (dra27) [inline manual/src/cmds/runtime.etex]: Is that right? I thought it’s 128 for 64-bit arches and 16 for 32-bit arches? (i.e. bytecode on 64-bit is max 128?) but I’m only looking on my phone!

Comment (ghost) [reply manual/src/cmds/runtime.etex]: David is right. Bytecode runtime on 64-bit platforms will default to 128.

Comment (kayceesrk) [reply manual/src/cmds/runtime.etex]: You are right. I read the definitions too fast.

File: runtime/caml/domain.h [hunk 2]
```diff
@@ -85,7 +87,7 @@ CAMLextern void (*caml_domain_initialize_hook)(void);
 CAMLextern void (*caml_domain_stop_hook)(void);
 CAMLextern void (*caml_domain_external_interrupt_hook)(void);
 
-CAMLextern void caml_init_domains(uintnat minor_heap_wsz);
+CAMLextern void caml_init_domains(uintnat minor_heap_wsz, uintnat max_domains);
```
Comment (gasche) [inline runtime/caml/domain.h]: Nit: I would maybe expect the other parameter order, as `max_domains` is the most important parameter of the two.

File: runtime/domain.c [hunk 3]
```diff
@@ -261,28 +261,28 @@ CAMLexport uintnat caml_minor_heaps_end;
 static CAMLthread_local dom_internal* domain_self;
 
 /*
- * This structure is protected by all_domains_lock
- * [0, participating_domains) are all the domains taking part in STW sections
- * [participating_domains, Max_domains) are all those domains free to be used
+ * This structure is protected by all_domains_lock.  [0, participating_domains)
+ * are all the domains taking part in STW sections.  [participating_domains,
+ * caml_params->max_domains) are all those domains free to be used.
```
Comment (gasche) [inline runtime/domain.c]: Nit: I would rather keep the `[foo, bar)` intervals on a single line each for readability, which requires doing some manual line-breaking here.

File: otherlibs/systhreads/st_stubs.c [hunk 4]
```diff
@@ -545,6 +546,18 @@ CAMLprim value caml_thread_initialize(value unit)
...
+                                       sizeof(struct caml_thread_table));
+  if (thread_table == NULL)
+    caml_fatal_error("caml_thread_initialize: failed to allocate thread"
+                     " table");
+
+  tick_thread_stop = caml_stat_alloc_noexc(caml_params->max_domains *
+                                           sizeof(atomic_uintnat));
+  if (tick_thread_stop == NULL)
+    caml_fatal_error("caml_thread_initialize: failed to allocate tick"
+                     " thread stop");
```
Comment (gasche) [inline otherlibs/systhreads/st_stubs.c]: Could we store `tick_thread_stop` in `thread_table` instead? This is not strictly-speaking related to the current PR, but it would simplify it noticeably and may improve other things in the future.

(In OCaml 4 we had a single global variable. The choice to make it a separate domain-indexed array was made very early in the threads-on-multicore implementation work, 7994082605daf9d39de55c57d5450e74c221838d ).

File: runtime/domain.c [hunk 5]
```diff
@@ -935,13 +936,29 @@ void caml_update_minor_heap_max(uintnat requested_wsz) {
...
+      caml_stat_alloc_noexc(max_domains * sizeof(dom_internal*));
+  if (stw_request.participating == NULL)
+    caml_fatal_error("Failed to allocate stw_request.participating");
+
+  stw_domains.domains =
+      caml_stat_alloc_noexc(max_domains * sizeof(dom_internal*));
+  if (stw_domains.domains == NULL)
+    caml_fatal_error("Failed to allocate stw_domains.domains");
+
+  caml_init_gc_stats(max_domains);
```
Comment (gasche) [inline runtime/domain.c]: I think this call should probably move to `caml_init_gc`, as it initializes a runtime service that is independent from the domains machinery. (I have no idea what `caml_init_signal_handling` is doing at the end of the current function, and the same comment would apply -- or maybe even move it to `caml_start_*` -- but oh well.)

File: runtime/gc_stats.c [hunk 6]
```diff
@@ -101,13 +102,20 @@ void caml_orphan_alloc_stats(caml_domain_state *domain) {
...
    all of these events could happen during domain termination. */
-static struct gc_stats sampled_gc_stats[Max_domains];
+static struct gc_stats* sampled_gc_stats;
+
+void caml_init_gc_stats (uintnat max_domains)
+{
+  sampled_gc_stats =
+    caml_stat_alloc(max_domains * sizeof(struct gc_stats));
+  if (sampled_gc_stats == NULL)
+    caml_fatal_error("Failed to allocate sampled_gc_stats");
```
Comment (gasche) [inline runtime/gc_stats.c:117]: I think that you meant `caml_stat_alloc_noexc`, as `caml_stat_alloc` fails with Out_of_memory instead of returning NULL. (I'm not sure if the hand-written init message is worth it, no strong opinion.) The other calls in the patch are all to `_noexc` so they are fine.

File: runtime/domain.c [hunk 7]
```diff
@@ -1724,7 +1741,9 @@ void caml_interrupt_self(void)
 /* async-signal-safe */
 void caml_interrupt_all_signal_safe(void)
 {
-  for (dom_internal *d = all_domains; d < &all_domains[Max_domains]; d++) {
+  for (dom_internal *d = all_domains;
+       d < &all_domains[caml_params->max_domains];
```
Comment (gadmm) [inline runtime/domain.c:1747]: This is signal-safe as (`all_domains` and) `caml_params->max_domains` is set before signal handlers are installed and it does not change afterwards. It might be good however to comment this fact.

File: runtime/domain.c [hunk 8]
```diff
@@ -935,13 +936,29 @@ void caml_update_minor_heap_max(uintnat requested_wsz) {
   check_minor_heap();
 }
 
-void caml_init_domains(uintnat minor_heap_wsz) {
+void caml_init_domains(uintnat minor_heap_wsz, uintnat max_domains) {
   int i;
 
+  all_domains = caml_stat_alloc_noexc(max_domains * sizeof(dom_internal));
```
Comment (gadmm) [inline runtime/domain.c]:
```suggestion
  all_domains = caml_stat_calloc_noexc(max_domains * sizeof(dom_internal));
```
Users inside `caml_try_run_on_all_domains_with_spin_work` and `caml_interrupt_all_signal_safe` expect it to be zero-initialised.

**Comment by kayceesrk** [comment]
Still seeing Windows CI failures after the bootstrap. @dra27 do you see other possibilities for failures that we see?

File: otherlibs/systhreads/st_stubs.c [hunk 9]
```diff
@@ -545,6 +546,18 @@ CAMLprim value caml_thread_initialize(value unit)
...
                   "while several domains are running.");
 
+  thread_table = caml_stat_alloc_noexc(caml_params->max_domains *
+                                       sizeof(struct caml_thread_table));
+  if (thread_table == NULL)
+    caml_fatal_error("caml_thread_initialize: failed to allocate thread"
+                     " table");
+
+  tick_thread_stop = caml_stat_alloc_noexc(caml_params->max_domains *
+                                           sizeof(atomic_uintnat));
```
Comment (gadmm) [inline otherlibs/systhreads/st_stubs.c]: As an alternative, perhaps simpler solution, these two variables implementing some domain-local state could be moved into dedicated Caml_state fields. It should not be a problem if people not using systhreads allocate two fields but do not use them.

File: runtime/gc_stats.c [hunk 10]
```diff
@@ -101,13 +102,20 @@ void caml_orphan_alloc_stats(caml_domain_state *domain) {
...
 }
 
-
 /* The "sampled stats" of a domain are a recent copy of its
    domain-local stats, accessed without synchronization and only
    updated ("sampled") during stop-the-world events -- each minor
    collection, major cycle (which potentially includes compaction),
    all of these events could happen during domain termination. */
-static struct gc_stats sampled_gc_stats[Max_domains];
+static struct gc_stats* sampled_gc_stats;
```
Comment (gadmm) [inline runtime/gc_stats.c:110]: Out of curiosity, is there a reason for not having this as a field in the domain state variable?

**Comment by gadmm** [comment]
> Boxroot is easily fixed by reading the new global instead, I believe. (This code runs inside the OCaml process, so it sees the same OCAMLRUNPARAM value).

It's not that easy, and the change is not ideal for us. `Max_domain` being known at compilation-time means that the array is allocated statically. This is used to implement an efficient domain-local state, which is relied-on for the fast path of boxroot. This usage of a fixed `Max_domain` has previously been discussed as we introduced the timing hook `caml_domain_terminated_hook` for this purpose (https://github.com/ocaml/ocaml/pull/11209). 

As previously discussed with @gasche, simply removing the limitation of `Max_domains` would remove an application without providing an alternative solution. Yet alternative implementations of DLS for C code would require (simple, but) more intrusive modifications to the runtime, and therefore need coordination with upstream prior to removal.

This PR presents an intermediate solution that does not remove Max_domains entirely, but makes it chosen at runtime. This changes the implementation of DLS as follows:
- One additional test for null when loading the DLS field.
- One additional indirection with a data depency on the previous test.
- This becomes a memory-inefficient way to implement a DLS when Max_domains is large (same issue for the present PR actually).

So this is not ideal for us, as the fast path of boxroot was carefully optimised (and we really gave advance notice about this usage of Max_domains).

Two possible solutions are not too complex to implement:
- More efficient DLS could be offered natively. For instance, another domain state variable that works like `Caml_state` could be introduced, with slots allocated dynamically like an implicit TLS). This also ought to be useful for the runtime, e.g. the two uses of DLS inside systhreads (except if you accept my alternative suggestion to reserve a `Caml_state` slot for them).
- A quick solution could be to please add one field for us (boxroot) inside the domain state. (A quick solution indeed although I don't like the idea of boxroot being priviledged compared to other libraries.)

**Comment by gadmm** [comment]
Thinking further, there might be a third solution to our problem of an efficient DLS, via caching inside an (efficient) implicit TLS variable. In this case, it might not matter very much if DLS is made a bit less efficient. I will think about it further and let you know if there are any issue in the end with the current PR.

File: runtime/gc_stats.c (hunk 10 above)
Comment (kayceesrk) [reply runtime/gc_stats.c:110]: There are a few more failure cases to handle if `sampled_gc_stats` were moved to `Caml_state`. In particular, `Caml_state` is created on demand when domains are created. If `gc_stats*` were a field in `Caml_state`, even with the guarantee that stats are only sampled in stop-the-world sections, we need to check for whether `gc_stats` is not NULL (since newly created domains may not participate in stw sections).

I tried to make this change, but the changes for handling failure cases got messy. Perhaps one solution is to have `sampled_gc_stats` in `dom_internal` struct. But that struct is internal to `domain.c`. 

Something to come back to later.

File: runtime/caml/domain.h (hunk 2 above)
Comment (kayceesrk) [reply runtime/caml/domain.h]: Done in https://github.com/ocaml/ocaml/pull/13272/commits/263be361baad53f6ef36f8774ac8b5d9b1572525.

File: runtime/domain.c (hunk 3 above)
Comment (kayceesrk) [reply runtime/domain.c]: Done in https://github.com/ocaml/ocaml/pull/13272/commits/263be361baad53f6ef36f8774ac8b5d9b1572525.

File: otherlibs/systhreads/st_stubs.c (hunk 4 above)
Comment (kayceesrk) [reply otherlibs/systhreads/st_stubs.c]: Done in https://github.com/ocaml/ocaml/pull/13272/commits/263be361baad53f6ef36f8774ac8b5d9b1572525.

File: runtime/domain.c (hunk 5 above)
Comment (kayceesrk) [reply runtime/domain.c]: Done in https://github.com/ocaml/ocaml/pull/13272/commits/263be361baad53f6ef36f8774ac8b5d9b1572525.

File: runtime/gc_stats.c (hunk 6 above)
Comment (kayceesrk) [reply runtime/gc_stats.c:117]: Done in https://github.com/ocaml/ocaml/pull/13272/commits/263be361baad53f6ef36f8774ac8b5d9b1572525.

File: runtime/domain.c (hunk 7 above)
Comment (kayceesrk) [reply runtime/domain.c:1747]: Done in https://github.com/ocaml/ocaml/pull/13272/commits/263be361baad53f6ef36f8774ac8b5d9b1572525.

File: runtime/domain.c (hunk 8 above)
Comment (kayceesrk) [reply runtime/domain.c]: Done in https://github.com/ocaml/ocaml/pull/13272/commits/263be361baad53f6ef36f8774ac8b5d9b1572525.

File: otherlibs/systhreads/st_stubs.c (hunk 9 above)
Comment (kayceesrk) [reply otherlibs/systhreads/st_stubs.c]: It seemed appropriate to move it to the `thread_table` in `st_stubs.c`. Done that in https://github.com/ocaml/ocaml/pull/13272/commits/263be361baad53f6ef36f8774ac8b5d9b1572525.

**Comment by kayceesrk** [comment]
> Still seeing Windows CI failures after the bootstrap. @dra27 do you see other possibilities for failures that we see?

Windows CI runs are successful now. Odd that they were failing earlier.

File: otherlibs/systhreads/st_stubs.c (hunk 9 above)
Comment (gadmm) [reply otherlibs/systhreads/st_stubs.c]: ok

File: runtime/gc_stats.c (hunk 10 above)
Comment (gadmm) [reply runtime/gc_stats.c:110]: ok, sounds like a good enough reason

File: otherlibs/systhreads/st_stubs.c (hunk 9 above)
Comment (gasche) [reply otherlibs/systhreads/st_stubs.c]: I have mixed feelings about this suggestion (for and against). This does not matter as I think you came to a satisfying resolution already, but let me elaborate in case it can help for future discussions.

1. I think that systhreads are there to stay, and that we should not hesitate to integrate them more in the runtime to make their implementation better, simpler, more correct.
2. On the other hand, I like that we still have code lying around in the runtime that behaves sort of like a third-party consumer of the runtime interfaces, instead of being hardcoded in every place. The *ability* for third-party libraries to have domain-local state as an array is useful, and keeping this aspect of the systhreads implementation is useful as a canary to check that we preserve this use-case. (@gadmm points out that the present change incurs a performance regression, which may matter for our current boxroot prototype; I don't know if it does matter for systhreads).

To conclude: if there are strong benefits to moving things into Caml_state, let's do it. But if we can do without, it gives us useful information about what third-party libraries can do, and there is value in that as well.

Comment (gasche) [reply otherlibs/systhreads/st_stubs.c]: (This reasoning somehow applies to `gc_stats` as well.)

Comment (gadmm) [reply otherlibs/systhreads/st_stubs.c]: That's a roundabout way of saying that the runtime could implement a proper DLS for itself because it would also benefit libraries.

**Review by gadmm — APPROVED** [review]
This looks good to me.

**Comment by gadmm** [comment]
Regarding the usage of `Max_domains` to implement a DLS, I am now pretty confident that we can get better performance by caching a pointer to DLS using implicit TLS, than any other possible implementation of DLS. Indeed, implicit TLS is backed by linker tricks that are otherwise inaccessible. (This makes the assumption that a system thread cannot change domains, but this assumption is probably deeply entrenched everywhere in the runtime implementation anyway.)

It is still useful to have a limit number of domains to implement the backing DLS, though. So it is a different story than if users started taking very large values for max_domains (currently limited to 4096) or if max_domains was removed entirely.

File: otherlibs/systhreads/st_stubs.c (hunk 9 above)
Comment (gasche) [reply otherlibs/systhreads/st_stubs.c]: Yep, "roundabout" is my middle name.

**Comment by kayceesrk** [comment]
> (This makes the assumption that a system thread cannot change domains, but this assumption is probably deeply entrenched everywhere in the runtime implementation anyway.)

Indeed. We have also documented this fact here: https://ocaml.org/manual/5.2/parallelism.html#s:par_systhread_interaction. 

> So it is a different story than if users started taking very large values for max_domains (currently limited to 4096)

We will have plenty of reasons to celebrate if someone usefully takes advantage of 4096 CPU cores in a single process. :-) 

Thanks for the approval. I'm debugging the Windows failures now.

**Comment by gasche** [comment]
@kayceesrk would it also be possible to have a Sandmark run for this PR? I don't expect any performance difference because I don't think we rely intensively on domain-indexed arrays in the runtime, but @gadmm comments made me curious about potential performance impact.

**Comment by gadmm** [comment]
I don't think it is necessary to benchmark the change. We're talking about microarchitectural effects here, and the code changed by this PR is not optimised that much.

File: runtime/caml/startup_aux.h [hunk 11]
```diff
@@ -52,6 +52,8 @@ struct caml_params {
   uintnat runtime_warnings;
   uintnat cleanup_on_exit;
   uintnat event_trace;
+
```
Comment (ghost) [inline runtime/caml/startup_aux.h]: No need for that extra newline here.

File: otherlibs/systhreads/st_stubs.c [hunk 12]
```diff
@@ -545,6 +550,12 @@ CAMLprim value caml_thread_initialize(value unit)
     caml_failwith("caml_thread_initialize: cannot initialize Thread "
                   "while several domains are running.");
 
+  thread_table = caml_stat_calloc_noexc(caml_params->max_domains,
+                                        sizeof(struct caml_thread_table));
+  if (thread_table == NULL)
+    caml_fatal_error("caml_thread_initialize: failed to allocate thread"
```
Comment (ghost) [inline otherlibs/systhreads/st_stubs.c:556]: At first I was concerned that the previous failure point in this function uses `caml_failwith`, but it can only occur in a specific cinematic where it makes sense to raise an exception, while these added allocation failures should be considered fatal (hence `caml_fatal_error`).

Eventually (in a later PR) this routine should be allowed to return a `caml_result`, but that's more work (and orthogonal to the problem being solved here).

File: runtime/domain.c [hunk 13]
```diff
@@ -1721,10 +1738,14 @@ void caml_interrupt_self(void)
   interrupt_domain_local(Caml_state);
 }
 
-/* async-signal-safe */
+/*  This function is signal-safe as [all_domains] and
+    [caml_params->max_domains] are set before signal handlers are installed and
+    it does not change afterwards. */
```
Comment (ghost) [inline runtime/domain.c]:
```suggestion
    their values do not change afterwards. */
```

**Review by ghost — APPROVED** [review]

File: otherlibs/systhreads/st_stubs.c (hunk 12 above)
Comment (gadmm) [reply otherlibs/systhreads/st_stubs.c:556]: Proper error handling in this function is actually part of #12410.

(Incidentally I noticed that this PR interferes with many others: #12410, #12964, #13026.)

**Comment by kayceesrk** [comment]
There is one last failure on MSVC 64-bits: https://github.com/ocaml/ocaml/actions/runs/9774768473/job/26994732983#step:13:1862. I'm struggling to set up a development environment to debug this.

**Comment by kayceesrk** [comment]
CI runs are successful now. It looks like not zero initialising what used to be statically allocated arrays was causing failures.

File: runtime/caml/startup_aux.h (hunk 11 above)
Comment (kayceesrk) [reply runtime/caml/startup_aux.h]: Fixed.

**Comment by kayceesrk** [comment]
CI is green now. We've had approvals from @gadmm and @dustanddreams. @gasche are we ok to merge?

**Review by gasche — APPROVED** [review]

**Comment by gasche** [comment]
@gadmm makes the point that this introduces conflicts with other in-flight PRs ( #12410, #12964, #13026 ). I still believe that "merge whichever is ready first" is a good principle, so I'm happy to merge this one, but I think we should also take this remark as a warning sign that we have a backlog of runtime PRs to handle, not to let fester for too long. I haven't been available to do much OCaml hacking in the last few weeks (and this will remain the case for another couple weeks). Clearly for #12410 the ball is in my camp and I will try to have a look.

File: otherlibs/systhreads/st_stubs.c (hunk 12 above)
Comment (gadmm) [reply otherlibs/systhreads/st_stubs.c:556]: Actually, I think these fatal errors should be `Out_of_memory` exceptions instead, since the `caml_failwith` above shows that it is admissible to fail with uninitialized thread machinery.

Now this raises the question of whether the `caml_failwith` above is sound. It is sound if we are sure that an uncaught exception during module initialization stops the program (otherwise we need hardening by checking the variable `threads_initialized` at the entry points of the Thread module). Unfortunately I don't know enough about module initialization to answer this question.

Comment (kayceesrk) [reply otherlibs/systhreads/st_stubs.c:556]: Is it a good idea to continue program execution when `caml_stat_calloc_noexc` fails to allocate memory for `thread_table`? Given that the underlying `calloc` has failed, the program is very likely doomed to fail soon in a non-recoverable way. Better to hard fail (fatal error) as soon as we anticipate this.

Like @gadmm, I also have the question of whether `caml_failwith` is sound. In particular, the exception that may be raised by this function is not handled at the call site,

https://github.com/ocaml/ocaml/blob/6ad156a735839b367fd2841bb3f6205316d9be4b/otherlibs/systhreads/thread.ml#L76-L79

and will lead to module initialisation failing with an uncaught exception?

Comment (gadmm) [reply otherlibs/systhreads/st_stubs.c:556]: My reasoning here is that the Out_of_memory will halt the program quickly anyway. Indeed, otherwise there is also an issue right above with `caml_failwith`. This can provide better debugging. Since this is a large allocation (e.g. can be above glibc MMAP_THRESHOLD), this does not mean that the remaining mallocs are going to fail.

But I agree that the main concern is the Failure above, which happens if another module spawns domains during initialization. Do we know if, when module initialization fails:
- no functions from this module can be called?
- the program stops?

Comment (gasche) [reply otherlibs/systhreads/st_stubs.c:556]: The semantics I *expect* if the evaluation of a module raises an exception is that the whole program fails with that exception, without executing the code of later modules in link order. (The mental model of `ocamlc a.ml b.ml c.ml` is that the produced program performs the sequential evaluation of modules A, B, C.)

However, taking a step back: I think this part of the runtime is an unfinished design, not something set in stone that we should think deeply about. We should admit that the runtime model exposed to our users is a small number of domains with arbitrarily many non-parallel threads per domain, and move whatever parts of the thread machinery into the `runtime/` logic that let us do this in a simpler, more robust way. It's "just" that no one around is actively working on this.

Comment (kayceesrk) [reply otherlibs/systhreads/st_stubs.c:556]: > Since this is a large allocation (e.g. can be above glibc MMAP_THRESHOLD), this does not mean that the remaining mallocs are going to fail.

Good point.

> We should admit that the runtime model exposed to our users is a small number of domains with arbitrarily many non-parallel threads per domain, and move whatever parts of the thread machinery into the runtime/ logic that let us do this in a simpler, more robust way.

Agree with this sentiment. It may help us move towards a cleaner runtime, avoiding the various hooks. 

I suspect that there will be a bit of design work that will be needed as the introduction of systhreads may add overheads in programs (due to the hooks). An easy first step may be to run Sandmark benchmarks with and without linking systhreads to see whether the overheads are significant.

**Comment by damiendoligez** [comment]
@gasche 
> why are we using one-letter name for OCAMLRUNPARAM variables, is this enforced by the specification or implementation? Otherwise I wouldn't mind `domains` or `max_domains` instead of `d`.

The parsing is done in C by a simple loop. We'd need a careful redesign (and more code) to switch to arbitrary identifiers.
